Kickstart elm/exercices from fsharp/exercises

mpizenberg opened the pull request:
This PR aims at preparing the Elm track for the upcomming work on concept exercises. I'll detail below what to do to finish this PR.

First, I need to design a minimalist `basics` or `minimal` example. It should aim at teaching the minimal concepts to create a compilable Elm module. The files required to modify are (for now):

 - [x] exercises/concept/basics/src/Basics.elm
 - [x] exercises/concept/basics/tests/Tests.elm

Once that concept exercise is ready, replace `src/Basics.elm` by a starter template, and move that implementation into `.meta/Example.elm`. Then, update the following document with the rationale for that exercise. It is directed to the people going to design the exercise (me for now).

 - [x] exercises/concept/basics/.meta/design.md

After that, update all the docs files:

 - [x] exercises/concept/basics/.docs/introduction.md (concepts/basics/about.md) # introduce concept
 - [x] exercises/concept/basics/.docs/instructions.md # exercise instructions
 - [x] exercises/concept/basics/.docs/hints.md        # help stuck students
 - [x] concepts/basics/links.json # reference documentation

Then, update the readme files common to all concept exercises

 - [x] exercises/concept/README.md
 - [x] reference/README.md
 - [x] config.json

Once that's done, we should update the guide to implement concept exercises

 - [x] reference/implementing-a-concept-exercise.md

Finally update some generic documents for the track:

 - [x] exercises/shared/.docs/cli.md

When all that is done the PR should be ready.

ErikSchierboom commented:
@mpizenberg Is there anything I can do to help with this PR?

mpizenberg commented:
@ErikSchierboom There is no real blocker except time spent on this. The fact is I have been busy for a few months basically rewriting an elm test runner to replace elm-test (what happens when we do `elm test`) because I was unhappy of how elm-test is able to work in the context of exercism for the online code editor. Probably looking for another person willing to help on the exercises front would be great but the elm community is quite small ...

ErikSchierboom commented:
@mpizenberg Okay thanks for the explanation! I'll keep an eye out for Elm enthusiasts that want to help.

ceddlyburge commented:
hi @mpizenberg , I'm going to start working on this now, so please let me know if you start doing things so we don't overlap.

mpizenberg commented:
Thanks Cedd! We won't overlap ^^.

ceddlyburge commented:
Hi @mpizenberg. Were you thinking of something really basic (probably just an import statement, an export statement and a function, like [hello](https://elm-lang.org/examples/hello)), or a very basic example of the elm architecture (like [increment / decrement a number(https://guide.elm-lang.org/)). 

I notice that we haven't got TEA on the list of concepts, so will need to add it if not doing it here. I am leaning towards just having the very basics in this exercise, and then slotting in TEA after "Import, Exposing, Functions, Float, Int, Type Annotations". I feel that if you were learning Elm, you would want to be able to create a function web app pretty early on in the process ....

mpizenberg commented:
>  like hello

Yes I was thinking of something like that, definitely not TEA. In my opinion, you need to understand a lot about elm before getting to something like TEA.

ceddlyburge commented:
Hi @mpizenberg, I've added the tests / code for a basic exercise. I've included `import` which isn't strictly necessary, but I think makes sense to include. Please let me know what you think, and then I'll move on to the supporting documentation (I think the plan for this has changed a bit since you made this PR, so I'll give everything a quick check first).

mpizenberg commented on languages/elm/exercises/concept/basics/src/Greet.elm:
```diff
@@ -0,0 +1,14 @@
+-- other modules (such as the tests) can use it
+module Greet exposing (greet)
+-- or
+-- module Basics exposing (..)
```
Still have "Basics" here. Maybe continue the previous explanation with something like `It is possible to export everything with "exposing (..)"` instead of the code example?

mpizenberg commented on languages/elm/exercises/concept/basics/src/Greet.elm:
```diff
@@ -0,0 +1,14 @@
+-- or
+-- module Basics exposing (..)
+
+-- you will need an `import` statement here to import the `toUpper` from the `String` module
```
the `toUpper` "function"

mpizenberg reviewed (commented):
Greet seems like a good entry point

ceddlyburge commented:
Hi @mpizenberg, I've added the remaining documentation for the exercise, can you take a look? And I'm confused about exercises/concept/basics/.docs/introduction.md and concepts/basics/about.md. I think about.md replaces introduction.md, but I couldn't find the pull request explaining the change, and the F# track has both, but they are basically the same.

ceddlyburge commented:
Hi @mpizenberg , I think this is all done now, can you give it a look? I'll then take a look at the failing github actions ...

ceddlyburge commented:
Hi @ErikSchierboom, I think this PR is ready now, can you give it a look (I can't get hold of @mpizenberg at the moment).

There are a coupld of failing github actions, but they don't appear to be related to changes in this pull request / things in the elm track, so maybe you could donate some advice on them.

Thanks, Cedd

ceddlyburge commented:
I seem to be having trouble getting hold of anyone, so I intend to just merge this myself. If anyone has an objection to this, please let me know soon!

kotp commented:
I could approve, but I don't think I am set for permissions here.  But hopefully the additional messaging will bump @ErikSchierboom  enough to look, and hopefully approve, or @mpizenberg .

If you think it is good though, probably reasonable to bring it in, and we could revert later, or update if there are changes needed.

mpizenberg commented:
Sorry for unresponsiveness. I'll have look during the weekend. Feel free to merge if you're confident, I can do a review after and eventually we can change problematic things if there are some.

ceddlyburge commented:
Happy to wait until the weekend. I'll start on the representer in the meantime ...

ErikSchierboom commented:
@mpizenberg Sorry, I don't work on Mondays and yesterday was a very hectic and busy day too. I'll take a look! Thanks for working on this.

ErikSchierboom commented on languages/elm/exercises/concept/basics/.docs/instructions.md:
```diff
@@ -0,0 +1,4 @@
+In this exercise you're going to write some minimal Elm code, to get a feel for the basics
+
+You should implement the `greet` function, to return "Welcome ", plus the `personToGreet` in upper case.
+
```
This document does not entirely conform to [the specs])(https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsinstructionsmd). For example, there is not really a story/theme here. Additionally, there should be individual tasks for each task the student should be doing (see [this example](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/lucians-luscious-lasagna/.docs/instructions.md)). Those tasks should include an example of calling the function to implement and its return value.

The exercise itself reminds me quite a bit of the existing `two-fer` exercise. Was there any reason you didn't copy any of the existing basics exercises? Like this [F# implementation](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/lucians-luscious-lasagna/.docs/instructions.md). One major benefit of re-using the existing exercise is that you can also re-use a lot of the existing documentation files. That said, you don't have to re-use an existing exercise of course, it's just that the `basics` concept exercise is a very tricky one to get right. For example, in this exercise the student is expected to work with strings, but the introduction doesn't actually introduce strings. For this reason, the other basics exercises all work with basic integers (no floats), as those will be familiar to anyone and don't require much introduction.

ceddlyburge replied:
The reason strings are used is that it requires you to use an `import`, which is part of what we defined as the basics. The string concatentation is in the hints, and in the test, but it could go in the introduction as well if you like?

`two-fer` in Elm uses a `Maybe` and potentially pattern matching, its actually reasonably far down the line in terms of Elm concepts.

I'm obviously not very keen to rewrite the exercise, but will leave the decision up to you. I think one issue for me is that the specs change faster than I can keep up with ...

ErikSchierboom commented on languages/elm/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,101 @@
+module Greet exposing (..)
+```
+
+https://elm-lang.org/docs/syntax#modules
````
The [introduction document specs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd) state:

> Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

I think it is best to remove it from this document. This also goes for the other links in this document.

ErikSchierboom commented on languages/elm/exercises/concept/basics/.docs/introduction.md:
```diff
@@ -0,0 +1,101 @@
+## Minimal compilable file
```
With the current spec, this should match the exercise's concept:

```suggestion
## basics
```

ErikSchierboom commented on languages/elm/exercises/concept/basics/.docs/introduction.md:
```diff
@@ -0,0 +1,101 @@
+
+### Indentation / significant whitespace
+
+Elm doesn’t use syntactic markers such as curly brackets, parentheses, or semicolons to specify code boundaries. It uses whitespaces and indentations instead.
```
```suggestion
Elm doesn't use syntactic markers such as curly brackets, parentheses, or semicolons to specify code boundaries. It uses whitespaces and indentations instead.
```

ErikSchierboom commented on languages/elm/exercises/concept/basics/.docs/introduction.md:
````diff
@@ -0,0 +1,101 @@
+-}
+```
+
+### Formatting
+
+There is a [style guide](https://elm-lang.org/docs/style-guide), and [elm-format](https://github.com/avh4/elm-format) can be used to automatically format code.
````
I think these should probably be removed here and added as hints. The above introduction looks like it allows the student to succesfully solve the exercise.

ErikSchierboom commented on languages/elm/exercises/concept/basics/src/Greet.elm:
```diff
@@ -0,0 +1,11 @@
+-- you will need an `import` statement here to import the `toUpper` function from the `String` module
+-- import String
+
+greet : String -> String
+greet personToGreet =
+    Debug.todo "Implement greet"
```
As the basics exercise introduces functions, one of its goals is to teach students how to define a function. We've found that the best way to do that is by having the students write the function from scratch. Most tracks "solve" this by having the stub file have some comment on what method to implement: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/lucians-luscious-lasagna/LuciansLusciousLasagna.fs#L3 Note that the basics exercise is rather exceptional, as virtually all other exercises will have stub methods in them (unless there is some inherent benefit of having students define something from scratch, e.g. when introducing the syntax for classes/structs/records).

ceddlyburge replied:
I don't think I agree with this, it just forces a student to copy paste a function from somewhere else. The syntax that a language uses to define a function is not the interesting thing to learn in my opinion. I've made the change anyway though, I just wanted to note my objection.

ErikSchierboom reviewed (commented):
I've left some comments. Sorry for the late review!

mpizenberg commented:
I've updated the basics concept to be the same (roughly) as the lucious lasagna F# example. This solves a couple of other discussed things here since we don't need to introduce the String module and module imports anymore. This can be done later on the concept exercise about strings. By the way @ErikSchierboom where is it already that we specify where the exercise comes from (since it comes from the F# track here)?

I've thus also updated the documentation introduction and concept readme to reflect the exercise changes. Let me know if all is good @ceddlyburge and @ErikSchierboom .

mpizenberg commented:
This warning pops up in the CI
```
Warning: elm/lucians-luscious-lasagna has no editor settings
```
but the rest do not seem to be elm related.

ErikSchierboom commented:
> By the way @ErikSchierboom where is it already that we specify where the exercise comes from (since it comes from the F# track here)?

This is done in the `.meta/config.json` file. [These docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#metaconfigjson) describe the format to use.

ErikSchierboom reviewed (approved):
This is great!
